src/core/net/dns_client.cpp

```diff
@@ -651,6 +651,22 @@ Error Client::ResolveAddress(const char *       aHostName,
     return StartQuery(info, aConfig, nullptr, aHostName, aContext);
 }
 
+#if OPENTHREAD_CONFIG_DNS_CLIENT_NAT64_ENABLE
+Error Client::ResolveIp4Address(const char *       aHostName,
+                                AddressCallback    aCallback,
+                                void *             aContext,
+                                const QueryConfig *aConfig)
+{
+    QueryInfo info;
+
+    info.Clear();
+    info.mQueryType                 = kIp4AddressQuery;
+    info.mCallback.mAddressCallback = aCallback;
+
+    return StartQuery(info, aConfig, nullptr, aHostName, aContext);
```

Comment:
Should we give up early if there is no NAT64 prefix available?

Reply:
The NAT64 prefix may be added by the time we get the response. Also if we check early and see we have NAT64 prefix, it may be removed by the time we get the response.

So makes sense to have a single place to check for it at the time we want to use it.

Reply:
> The NAT64 prefix may be added by the time we get the response. Also if we check early and see we have NAT64 prefix, it may be removed by the time we get the response.

Seems very unlikely that NAT64 prefix would change in such a short period of time?

Reply:
We cannot assume that it won't change (even if it is unlikely). So we need to check at point of use. :)

Reply:
An `A` record query from DNS client incurs Thread traffic and likely mDNS multicast query and response. 
Better to save the bandwidth if we know in advance that it's highly likely going to be a waste?

Reply:
I refactored the code moved the code which selects the nat64 prefix to `NetworkData::Leader`) and added such a check at start of query in the latest push.

I agree that it seems reasonable to add such check and optimization (avoid traffic if we can) but at the same time we need to be mindful about the slow impact of such small type of extra optimizations on the code size and be selective about them.